### ISSUE TITLE
Glitchy behavior when calling fitBounds multiple times

### ISSUE THREAD

**Issue by uniphil** [Leaflet/Leaflet#3249]
Demo: [0.7.3](http://codepen.io/uniphil/pen/zxjprZ), [master](http://codepen.io/uniphil/pen/NPMBqM).

Triggering code:

```javascript
map.fitBounds([[-20, -20], [20, 20]]);
map.fitBounds([[-10, -10], [10, 10]]);
```

**Expected behaviour**: final state of the map reflects the last call to `fitBounds`.

**Observed behaviour**: final state of the map might reflect either of the `fitBounds` calls, depending on the initial map state.

The effect is reproducible as long as the second `fitBounds` is called before the first has finished animating, eg:

```javascript
map.fitBounds([[-20, -20], [20, 20]]);
setTimeout(function() {
  map.fitBounds([[-10, -10], [10, 10]]);
}, 200);
```
or a user clicking two buttons in quick succession will also show the effect.

I don't have a simple testcase for it, but in my project, two quick `fitBounds` calls can also leave the map in weirder states (zoomed out very far) that don't reflect any of the `fitBounds` calls.

**Comment by mourner** [comment]
@uniphil hmm, so this is triggered only when both `fitBounds` calls have bounds with the same center, right?

**Comment by uniphil** [comment]
@mourner thanks for the quick response!

No, they can have different centres (the linked demos have different centres). However, it's only glitchy if a zoom occurs. [changing the demo so that `fitBounds` only causes a pan](http://codepen.io/uniphil/pen/vEjzrb) produces the expected behavior though:

```javascript
map.fitBounds([[-20, -20], [20, 20]]);
map.fitBounds([[0, 0], [40, 40]]);
```

so it seems to be something about the zoom, specifically.

**Comment by mourner** [comment]
I'm just wondering why it's specifically limited to `fitBounds` when it just calls `setView` under the hood...

**Comment by uniphil** [comment]
ah, yeah, probably because I didn't change the zoom level in my `setView` test :)

**Comment by uniphil** [comment]
[glitchy setView](http://codepen.io/uniphil/pen/LEmJgJ)

**Comment by mourner** [comment]
@perliedman I wonder why this reappeared after #2021...

**Comment by perliedman** [comment]
@mourner while they look similar, do you think it's really the same issue? #2021 occured because the transition end event was never fired unless `requestAnimFrame` was called, causing the map to appear locked up, since drag handlers etc. are disabled during zoom (just recalling from memory, so sorry if some details are incorrect).

With this issue, the map appears to continue to work as expected, but the location and zoom is not (always) the expected. So it has the looks of a race condition, but possibly some other sort of race condition.

Anyway., I'll try to look in to it a bit.

**Comment by perliedman** [comment]
Some notes after inspecting this problem:
* The same problem is reproducible with Leaflet 0.6.0, so it's been there for a long time
* When animation is enabled, the map's zoom (the value returned by `getZoom`) isn't updated until the zoom animation finishes
* If a zoom animation should be triggered or not is determined by comparing the target zoom level with the current zoom level
* If we're at zoom level `a`, two consecutive calls to change zoom levels, one to `a`, and one to `b`, will *always* initiate a zoom to `b`, no matter the order of the calls, since the change to zoom `b` will initiate an animated zoom, setting the zoom to `b` when it's finished, while the zoom to `a` will not initiate an animated zoom, since the map is already at this level

I think the last point explains the behavior we're seeing.

I guess the solution is to somehow abort a requested zoom animation if a new call comes in that sets the zoom level to the current zoom level, before the animation has finished (or, as in this case, in fact before it's even started because of the delay by `requestAnimFrame`).

Hope this explanation makes some sense.

**Comment by leoromanovsky** [comment]
I've seen this bug manifest as well. Some observations, if it helps:

* Turning off animation fixes the issue.

To add something new: besides markers, I'm also drawing a polyline on the map, all with GeoJSON. When calling setView with animation, the polyline ends up "stuck" in the old zoom, and often times shifted over.

This also is resolved without animation.

**Comment by vvo** [comment]
Also seeing this, I used animate: false and it's now ok.

Seems like this will be fixed in 1.1?

**Comment by theashyster** [comment]
Before discovering this issue I created a test case which I wanted to use to submit a new issue.
Adding it here instead http://playground-leaflet.rhcloud.com/recu/edit?html,output

**Comment by Jmuccigr** [comment]
I may be seeing something related: fitBounds will animate only if there's no zoom involved.

I first use fitBounds to set the map's bounds. Then if I pan and run that same fitBounds function, `animate: true` works and leaflet honors the duration value. If I just zoom, or add a zoom to the pan, fitBounds results in a very fast move to the bounding area, which will appear to be a simple switch to it, unless the change from the bounds area is fairly small, in which case you can see the move take place very quickly. (I assume this is because the tiles are cached or something, so leaflet can play them back as the quick move takes place.)

**Comment by Jmuccigr** [comment]
I'm still seeing my issue with 1.2.0.

**Comment by perliedman** [comment]
@Jmuccigr that is the expected behaviour. `fitBounds` will only animate if the point is close on the same zoom level (https://github.com/Leaflet/Leaflet/blob/master/src/map/Map.js#L1520) or the zoom level is within a threshold (https://github.com/Leaflet/Leaflet/blob/master/src/map/Map.js#L1575); also see the option [`zoomAnimationThreshold`] to control this.

If you always want an animation, you can use `flyToBounds` instead.

**Comment by perliedman** [comment]
The original issue is still reproducible in @theashyster's playground example. I think [my explanation here](https://github.com/Leaflet/Leaflet/issues/3249#issuecomment-75931374) is still a valid description of why this happens, but we need to figure out how.

**Comment by Jmuccigr** [comment]
@perliedman, thanks. I hadn't realized that. Is this something that might usefully go into the docs?

**Comment by atlantageek** [comment]
This is still open. Just a suggestion but can we have fitBounds and setZoom return a promise or callback when the animation is complete?

**Comment by perliedman** [comment]
@atlantageek I guess we theoretically could but

1. we don't use a lot of callbacks for async stuff in other places, so it is sort of against the Leaflet style
2. it would not fix this issue (as far as I understand)

If you really would like something like that, I suggest opening a separate issue with this suggestion. I do not want to sound harsh, but I would say it is not likely to be implemented unless you can argue for it and are prepared to implement it; Leaflet is quite mature and changing core APIs and/or patterns is not really on any maintainers todo at the moment.

**Comment by bygiro** [comment]
This bug is not resolved yet. The only solution, which worked for me, was to disable the animation.

**Comment by iByteDev** [comment]
Hello @bygiro , 
I have the same case, please could you tell me exactly how you solved it?

`var map = L.map('map), {
center: L.latLng(LATITUD_DEFAULT, LONGITUD_DEFAULT),
zoomControl: false,
minZoom: 4.50,
animate: false
}`

My map sometimes appears centered on a place where I am not setting the center.

**Comment by syonfox** [comment]
Ok so just looking into an issue on my project and I believe this is linked.  Might get into it and hack a fix if nobody has yet...

move then zoom vector animation is broken (order matter it seems)

this is reproducible here
https://github.com/Leaflet/Leaflet/issues/3249
https://leafletjs.com/examples/geojson/

Steps to reproduce:

Mouse down and start panning

Then while panning  zoom with scroll wheel (or second touch)

This is not so noticeable on desktop but on mobile

Touch down start panning first then touch with second finger and zoom
this happens quite often win users are just moving around with 2 fingers an is more prominent with continuous zoom.

Looks like something is not lining up.  @bygiro's solution does not work on mobile  bot does mitigate desktop all be it at a cost.

The fact that it works properly on mobile if you start to zoom first and ten pan might hint at some solutions

Idea time: 

Have the pan action with one finger Emulate 2 touches moving in parallel

Fix the second touch event to properly update the initial position by ending the pan before starting.

Anyone whos been in the guts have inputs? does this sound like its on the right track if so let me know.

**Comment by syonfox** [comment]
Ok so I did some hacking this seems to do the trick  I'm not sure how reliable it is.  The animation still messes up a little bit on mobile but its is more usable and on desktop it works perfect with scroll.

This fixes the above problem by ending the drag pan action before leaflet does the two finger drag or scroll zoom.

```js

let mapDiv = document.getElementById('map')
L.DomEvent.on(mapDiv, 'wheel', (e)=>{
    console.log('wheel')
    if(map.dragging.moving()) {//if its moving
        map.dragging._onDragEnd(e); //end the move action before leaflet handles the scroll
    }
});

L.DomEvent.on(mapDiv, 'touchstart', (e)=>{
    //https://github.com/Leaflet/Leaflet/blob/69f73ed1e8a83724c086ce206e7f9d7babc5d3dc/src/map/handler/Map.TouchZoom.js#L42
    if (!e.touches || e.touches.length !== 2 || map._animatingZoom /*|| this._zooming*/) { return; }
    console.log('touchstart')

    if(map.dragging.moving()) {
        map.dragging._onDragEnd(e);
    }
});
//...initialize map after I believe this makes it so that our event handlers get executed first.
```

**Comment by natevw** [comment]
Is the core issue here that Leaflet doesn't cancel previous animations when a new one is triggered? I've noticed a situation I think is related, where I am updating the map bounds as several layers each asynchronously load.

Basically something like:

```
let layer1, layer2;
loadData('layer_one').then((d) => {
  layer1 = addData(d, map);
  updateBounds();
});
loadData('layer_two').then((d) => {
  layer2 = addData(d, map);
  updateBounds();
});
function updateBounds() {
  if (layer1) map.setView(layer1.getBounds().getCenter(), 12);
  else if (layer2) map.fitBounds(layer2.getBounds());
}
```

Admittedly a little weird logic, but regardless, what I'm seeing is that by default I don't always get the expected boundaries once both layers have loaded. That is, the map should always be zoomed to the `layer1` state once both layers load. But if layer2 happens to load first I do not get that result.

If I add `{animate:false}` to both calls then it works as expected. So without diving into the code, my theory is that if the `map.fitBounds(layer2.getBounds())` animation is still in progress when I request `map.setView(layer1.getBounds().getCenter(), 12)` the earlier animation keeps progressing **too**, in parallel with and throwing off the end result of the new animation, perhaps?

**Comment by heyman** [comment]
> Is the core issue here that Leaflet doesn't cancel previous animations when a new one is triggered?

The following code seems to indicate that a previous animation *is* canceled when a new animation start:
https://github.com/Leaflet/Leaflet/blob/bbd264704b05036a281bd44513cf1760d8336b9d/src/map/Map.js#L174-L180
https://github.com/Leaflet/Leaflet/blob/bbd264704b05036a281bd44513cf1760d8336b9d/src/map/Map.js#L354-L361
https://github.com/Leaflet/Leaflet/blob/bbd264704b05036a281bd44513cf1760d8336b9d/src/map/Map.js#L1259-L1265

**Comment by heyman** [comment]
I'm running into this or a similar issue when using `Map.flyTo`/`Map.flyToBounds`. I have a page where the main content is on the left side of the screen, while the right side shows a map. Depending on where what's in the viewport in the left side, I use `flyTo` to highlight the relevant area on the map. This works great when browsing the page "slowly", but if you scroll the page fast (e.g. to get to the top of the page) it'll trigger multiple consecutive calls to `flyTo`.

I've found two workarounds that both fixes the glitchy behaviour:

1. Calling `map.fire('viewreset')` just before calling `flyTo`.
2. Setting [zoomSnap](https://leafletjs.com/reference.html#map-zoomsnap) to `0` and calling `map.stop()` just before calling `flyTo`.

So adding a line that fires the `viewreset` event to either `Map.flyTo` or `Map._stop` should fix the issue, I believe. However, I'm not familiar enough with the internal Leaflet code to know if that would cause any other issues?

**Comment by TadeoBennett** [comment]
Hello. The fix on mine was to call fitBounds() after the map has been rendered on the page by using a set timeout function for 200ms and turning animations off. My implementation here is a snipper of code from a vueJS application so my bounds are saved in a ref. 

```
     // used to reshow the hidden leaflet map after performing a request.
      pricesLoading.value = false;
      loadingPricesMap.value = false //hide the spinner
      showPricesMap.value = true //show the map

      setTimeout(() => {
        dynamicMap2.value.invalidateSize(); //make sure the map resizes correctly
      }, 100);

      setTimeout(() => {
        if (hadSetBounds.value) {
          dynamicMap2.value.fitBounds(bounds.value, {padding: [60, 60]});
          hadSetBounds.value = false;
          bounds.value = null;
        }
      }, 200);
    });
```

**Comment by axxjrl** [comment]
Fix for my React project was:
```
useEffect(() => {
        const bounds = getBoundsFromCollection<GeoJSON>(refs.reduce<GeoJSON[]>(
            (acc, ref) => {
                if (ref.current) {
                    return [...acc, ref.current];
                } else {
                    return acc;
                }
            }, []
        ));

        // https://github.com/Leaflet/Leaflet/issues/3249
        setTimeout(() => {
            map.invalidateSize();
        }, 100);

        setTimeout(() => {
            if (bounds && bounds.isValid()) {
                map.fitBounds(bounds, {padding: [60, 60]});
            }
        }, 200);
    }, [refs]);
```
